framework/encode/d3d12_capture_manager.cpp

```diff
@@ -1770,21 +1851,156 @@ void D3D12CaptureManager::PreProcess_ID3D12CommandQueue_ExecuteCommandLists(ID3D
+    {
+        auto trim_draw_calls = GetTrimDrawCalls();
+        // TODO: When queue_submit_count_ becomes 0-based, remove "-1".
+        if ((common_manager_->GetQueueSubmitCount() - 1) == trim_draw_calls.submit_index)
+        {
+            auto                            target_cmdlist = lists[trim_draw_calls.command_index];
+            ID3D12CommandList_Wrapper*      target_wrapper = nullptr;
+            target_cmdlist->QueryInterface(IID_IUnknown_Wrapper, reinterpret_cast<void**>(&target_wrapper));
+            GFXRECON_ASSERT(target_wrapper);
+
+            auto target_info = target_wrapper->GetObjectInfo();
+            GFXRECON_ASSERT(target_info);
+
+            if (!target_info->is_split_commandlist)
+            {
+                DecrementCallScope();
+                // target of splitted
+                common_manager_->ActivateTrimmingDrawCalls(format::ApiFamilyId::ApiFamily_D3D12);
+
+                std::vector<ID3D12CommandList*> cmdlists;
+
+                auto target_draw_call_cmd = target_info->split_command_sets[graphics::dx12::kDrawCallArrayIndex].list;
+                GFXRECON_ASSERT(target_draw_call_cmd);
+                cmdlists.emplace_back(target_draw_call_cmd);
+
+                wrapper->ExecuteCommandLists(cmdlists.size(), cmdlists.data());
+
+                cmdlists.clear();
+
+                common_manager_->DeactivateTrimmingDrawCalls();
+
+                // after of splitted and lists
+                auto after_draw_call_cmd =
+                    target_info->split_command_sets[graphics::dx12::kAfterDrawCallArrayIndex].list;
+                GFXRECON_ASSERT(after_draw_call_cmd);
+                cmdlists.emplace_back(after_draw_call_cmd);
+
+                for (uint32_t i = (trim_draw_calls.command_index + 1); i < num_lists; ++i)
+                {
+                    cmdlists.emplace_back(lists[i]);
+                }
+
+                wrapper->ExecuteCommandLists(cmdlists.size(), cmdlists.data());
+                is_complete = true;
+                IncrementCallScope();
+            }
+        }
+    }
+
+    if (!is_complete)
+    {
+        auto queue         = wrapper->GetWrappedObjectAs<ID3D12CommandQueue>();
+        auto unwrap_memory = GetHandleUnwrapMemory();
+        queue->ExecuteCommandLists(num_lists, UnwrapObjects<ID3D12CommandList>(lists, num_lists, unwrap_memory));
+    }
 }
 
 void D3D12CaptureManager::PostProcess_ID3D12CommandQueue_ExecuteCommandLists(ID3D12CommandQueue_Wrapper* wrapper,
                                                                              UINT                        num_lists,
                                                                              ID3D12CommandList* const*   lists)
 {
-    PostQueueSubmit();
+    GFXRECON_ASSERT(num_lists > 0 && lists);
+
+    ID3D12CommandList_Wrapper* cmd_wrapper = nullptr;
+    lists[0]->QueryInterface(IID_IUnknown_Wrapper, reinterpret_cast<void**>(&cmd_wrapper));
```

Comment:
I think this might return a pointer to an `IUnknown*` (not a `ID3D12CommandList_Wrapper*`) because  `call scope >= 1` and `IID_IUnknown_Wrapper` is used.

`ID3D12CommandList* const* lists` should already be wrapper pointers, they just need to be reinterpret cast:
`ID3D12CommandList_Wrapper* cmd_wrapper = reinterpret_cast<ID3D12CommandList_Wrapper>(list[0]);`

This also applies to other places in this PR. Generally `QueryInterface` should not be used to get a wrapper object.

Reply:
I don't know. When I used `reinterpret_cast` to cast to wrapper, I sometimes got nullptr. I didn't study deeply why it happened. But `QueryInterface` never happened nullptr issue.  I feel `QueryInterface` is safer. I could try `reinterpret_cast` again to see if nullptr happens again.

Reply:
fixed

Reply:
> When I used reinterpret_cast to cast to wrapper, I sometimes got nullptr.
Does this still happen? I wouldn't expect any nullptrs there. For example `Dx12StateTracker::TrackExecuteCommandLists` does the following and doesn't null-check the wrapper.
```
        auto list_wrapper = reinterpret_cast<ID3D12CommandList_Wrapper*>(lists[i]);
        auto list_info    = list_wrapper->GetObjectInfo();
```

Reply:
I changed to `reinterpret_cast` and didn't see nullptr happen again. I think I made mistakes to use `reinterpret_cast` before. Maybe I tried casting a native object to a wrapper.